apps/trading/components/welcome-dialog/welcome-dialog.tsx

```diff
@@ -24,39 +25,58 @@ export const WelcomeDialog = () => {
   const { pathname } = useLocation();
   const { VEGA_ENV } = useEnvironment();
   const [dialog, setDialog] = useState<DialogConfig | null>(null);
-  const onClose = useCallback(() => {
-    setDialog(null);
-  }, [setDialog]);
   const [riskAccepted] = useLocalStorage(constants.RISK_ACCEPTED_KEY);
-
   const { data } = useDataProvider({
     dataProvider: activeMarketsProvider,
   });
 
-  useMemo(() => {
+  const { update, marketId, shouldDisplayWelcomeDialog } = useGlobalStore(
+    (store) => ({
+      update: store.update,
+      marketId: store.marketId,
+      shouldDisplayWelcomeDialog: store.shouldDisplayWelcomeDialog,
+    })
+  );
+
+  const isRiskDialogNeeded =
+    riskAccepted !== 'true' && VEGA_ENV === Networks.MAINNET;
+  const isWelcomeDialogNeeded =
+    (pathname === '/' && !marketId) || shouldDisplayWelcomeDialog;
+  const onClose = useCallback(() => {
+    update({ shouldDisplayWelcomeDialog: isRiskDialogNeeded });
+    setDialog(null);
+  }, [setDialog, update, isRiskDialogNeeded]);
+
+  useEffect(() => {
```

Comment:
suggestion (non blocking): I realize this was already like this.. but hy are we setting components to state here? Can we just have some render logic? I think a small refactor might be good to simplify, increase readability and reduce renders. Something like:

```
// in component

const WelcomeDialog = () => {
  ...
  let dialog = null
  
  
  if (riskAccepted !== 'true' && VEGA_ENV === Networks.MAINNET) {
    dialog = {
      content: <WelcomeNoticeDialog />,
      title: t('WARNING'),
      size: 'medium'
    }
  } else if (pathname === '/' && data?.length === 0) {
    ...
  } else if (etc etc) {
    ...
  }
  
  return (
    <Dialog
      open={dialog !== null}
      title={dialog.title}
      size={dialog.size}
      onChange={onClose}
    >
      {dialog !== null && dialog.content}
    </Dialog>
  )  
}
```

Reply:
We rely here on routing parameter (_pathname_). If we gave up using _state_ here and started to simply return rendered component, changing pathname after a redirect would change the conditions and dialog could be hidden.
The goal was to avoid using store flag. Finally I had to use the store flag here due to growing complexity, so maybe I'll change the implementation, but still simply returning rendered content will be not easily feasible, because we will depend on store flag then  ¯\_(ツ)_/¯